README.md
### Language(语言选择)
English     [中文](./README-zh_CN.md)

###  Table Of Contents
+ [Instruction](#intro)
+ [Installation](#install)
+ [Usage](#use)


### Instruction
`markdown-clear` is a NodeJS tool to automatic order the markdown headers to make it clear to read,especially on long document.
+ example
```
#     ->    # 1
##    ->    ## 1.1
##    ->    ## 1.2
###   ->    ### 1.2.1
###   ->    ### 1.2.2
#     ->    # 2
##    ->    ## 2.1
```
+ you are not forced to use h1,but with the most stronger head on the first.
```
###     ->    ### 1
####    ->    #### 1.1
####    ->    #### 1.2
#####   ->    ##### 1.2.1
###     ->    ### 2
####    ->    #### 2.1
```

### Installation
use npm:
```
npm install markdown-clear -g
```
or use yarn
```
yarn global add markdown-clear
```

### Usage
+ transform a single file
```
markdown-clear test.md test-out.md
```
+ update the content of the same file
```
markdown-clear test.md --force
```

+ transform a directory
```
markdown-clear test-dir test-dir-out
```

+ transform a directory and it's sub-directories
```
markdown-clear test-dir test-dir-out -r
```

+ update files in a directory
```
markdown-clear test-dir --force
```

+ update files in a directory and it's sub-directories
```
markdown-clear test-dir --force -r
```

CHANGELOG


### 1.0.1
+ support auto order the valid headers

tests/index.js
/**
 * --------------------------------------------------------------------------- *
 *
 * @Project: markdown-clear
 * @FileName: index.js
 * @Dependence: --
 * @Description: --
 * @CreatedBy: liao.zhang
 * @CreateDate: 2017/8/3 23:20
 * @LastModifiedBy: liao.zhang
 * @LastModifiedDate: 2017/8/3 23:20
 *
 * --------------------------------------------------------------------------- *
 */

const test = require('../../markdown-clear')

tests/utils/FSUtil.js
/**
 * --------------------------------------------------------------------------- *
 *
 * @Project: markdown-clear
 * @FileName: FSUtil
 * @Dependence: --
 * @Description: --
 * @CreatedBy: liao.zhang
 * @CreateDate: 2017/7/30 16:25
 * @LastModifiedBy: liao.zhang
 * @LastModifiedDate: 2017/7/30 16:25
 *
 * --------------------------------------------------------------------------- *
 */

const {FsUtil} = require('../../lib/utils/FSUtil');

console.log('1111:',FsUtil.getPathList('../../',['.md','.markdown']));

console.log('2222:',FsUtil.getPathList('../../',['.md','.markdown'],['.git','node_modules']));

console.log('3333:',FsUtil.getPathList('../../src',['.js'],['.js']));



tests/modules/HeaderOrder.js
/**
 * --------------------------------------------------------------------------- *
 *
 * @Project: markdown-clear
 * @FileName: HeaderOrder.js
 * @Dependence: --
 * @Description: --
 * @CreatedBy: liao.zhang
 * @CreateDate: 2017/8/2 22:48
 * @LastModifiedBy: liao.zhang
 * @LastModifiedDate: 2017/8/2 22:48
 *
 * --------------------------------------------------------------------------- *
 */

const {HeaderOrder}  = require('../../lib/modules/HeaderOrder');

HeaderOrder.processContent(`
### aaa
#### bb
`);

src/cli.js
#!/usr/bin/env node

/**
 * --------------------------------------------------------------------------- *
 *
 * @Project: markdown-clear
 * @FileName: cli.js
 * @Dependence: --
 * @Description: --
 * @CreatedBy: liao.zhang
 * @CreateDate: 2017/8/3 19:48
 * @LastModifiedBy: liao.zhang
 * @LastModifiedDate: 2017/8/3 19:48
 *
 * --------------------------------------------------------------------------- *
 */

const fs = require('fs');
const {FsUtil} = require('./utils/FSUtil');
const {HeaderOrder}  = require('./modules/HeaderOrder');

const yargs = require('yargs');
const argsObj = yargs
    .usage('Usage: $0 input [output] [-r]')
    .help('h')
    .option('r',{
        alias:'recursive',
        boolean:true,
        describe:'recursive process markdown files in directory'
    })
    .option('force',{
        boolean:true,
        describe:'force to update on the origin file'
    })
    .argv;
const argsList = argsObj._;

if(argsList.length === 0){
    yargs.showHelp();
}else{
    let input = argsList[0],output = argsList[1];
    if(!output && !argsObj.force){
        yargs.showHelp();
    }else{
        if(argsObj.force){
            output = input;
        }

        fs.stat(input,(err,stats) => {
            if(err){
                throw err;
            }else{
                if(stats.isFile()){
                    HeaderOrder.processFile(input,output);
                }else if(stats.isDirectory()){
                    HeaderOrder.processDir(input,output,argsObj.recursive);
                }else{
                    //noop
                }
            }
        })
    }

}

src/utils/FSUtil.js
/**
 * --------------------------------------------------------------------------- *
 *
 * @Project: markdown-clear
 * @FileName: FSUtil
 * @Dependence: --
 * @Description: --
 * @CreatedBy: liao.zhang
 * @CreateDate: 2017/7/30 16:17
 * @LastModifiedBy: liao.zhang
 * @LastModifiedDate: 2017/7/30 16:17
 *
 * --------------------------------------------------------------------------- *
 */

import sysFs from 'fs'
import sysPath from 'path'
import mkdirp from 'mkdirp'

const FsUtil = {
    getPathList: (basePath = './', fileExt = '',ignore = [], recursive = true) => {
        const pathList = [];
        let extList = [],ignoreList = [];
        if(fileExt){
            if(typeof fileExt === 'string'){
                extList.push(fileExt)
            }else if(Array.isArray(fileExt)){
                extList = fileExt.map((item) => {
                    return item.toLowerCase();
                });
            }else{
                //noop
            }
        }

        if(ignore){
            if(typeof ignore === 'string'){
                ignoreList.push(ignore)
            }else if(Array.isArray(ignore)){
                ignoreList = ignore;
            }
        }

        FsUtil._fillPathList(pathList,basePath, extList, ignoreList, recursive);

        return pathList;
    },
    _getPathStat: (path) => {
        let pathStat;
        try {
            pathStat = sysFs.statSync(path);
        } catch (e) {
            pathStat = null;
            console.log('获取路径状态失败:', path, '\n', e)
        }
        return pathStat;
    },
    _fillPathList: (fileList,basePath,extList,ignoreList, recursive) => {
        const self = FsUtil;
        if(!ignoreList.includes(sysPath.basename(basePath))){
            let pathStat = self._getPathStat(basePath);
            if (pathStat) {
                if (pathStat.isFile()) {
                    if (extList.includes(sysPath.extname(basePath).toLowerCase())) {
                        fileList.push(basePath);
                    }
                } else if (pathStat.isDirectory()) {
                    const files = sysFs.readdirSync(basePath);
                    files.forEach(function (name) {
                        const nextPath = sysPath.resolve(basePath, name);
                        pathStat = self._getPathStat(nextPath);
                        if (pathStat) {
                            if (pathStat.isFile()) {
                                if (extList.includes(sysPath.extname(nextPath).toLowerCase())) {
                                    fileList.push(nextPath);
                                }
                            } else if (pathStat.isDirectory()) {
                                recursive && self._fillPathList(fileList,nextPath,extList,ignoreList, recursive);
                            } else {
                                //noop
                            }
                        }
                    })
                } else {
                    //noop
                }
            }
        }
    },
    mkdirp:(path) => {
        mkdirp.sync(path)
    }
}


export default FsUtil
export {
    FsUtil
}

src/modules/HeaderOrder.js
/**
 * --------------------------------------------------------------------------- *
 *
 * @Project: markdown-clear
 * @FileName: HeaderOrder
 * @Dependence: --
 * @Description: --
 * @CreatedBy: liao.zhang
 * @CreateDate: 2017/7/30 16:22
 * @LastModifiedBy: liao.zhang
 * @LastModifiedDate: 2017/7/30 16:22
 *
 * --------------------------------------------------------------------------- *
 */

import sysFs from 'fs'
import sysPath from 'path'
const {FsUtil} = require('../utils/FSUtil');

const LINE_SEP = '\n';
const HEADER_LEN = 6;

const HeaderOrder = {
    processDir:(srcPath,dstPath,recursive,encoding='utf-8') => {
        const fileList = FsUtil.getPathList(srcPath,['.md','.markdown'],['.git','node_modules'],recursive);
        const baseSrcPath = sysPath.resolve(srcPath);
        const baseDstPath = sysPath.resolve(dstPath);
        fileList.forEach((path) => {
            HeaderOrder.processFile(path,path.replace(baseSrcPath,baseDstPath));
        });
    },
    processFile:(srcPath,dstPath,encoding='utf-8') => {
        sysFs.stat(srcPath,(err,stats) => {
            if(err){
                throw err;
            }

            const self = HeaderOrder;
            sysFs.readFile(srcPath,encoding,(err,data) => {
                const newContent = self.processContent(data);
                const dstDir = sysPath.dirname(dstPath);
                if(!sysFs.existsSync(dstDir)){
                    FsUtil.mkdirp(dstDir);
                }
                sysFs.writeFile(dstPath,newContent,'utf-8',(err) => {
                    if(err){
                        throw err
                    }
                })
            });
        })
    },
    processContent:(content) => {
        const fileLines = content.split(LINE_SEP);
        const headerRegex = /^(#+)\s*(.*)\s*(#?)/;
        const orderedHeaderRegex = /^(#+)\s*(\d+([\.\d+]+)?\.?)\s*(#+)?/;

        const firstHeader = HeaderOrder.findFirstHeader(fileLines);
        const headerList = [0,0,0,0,0,0];// h1,h2...h6
        if(firstHeader){
            headerList.splice(0, firstHeader.length - 1);
        }

        let tmpItem,tmpHeader,tmpHeaderOrder;
        const resultLines = fileLines.map((item) => {
            //如果已经有Header 序号就清空
            tmpItem = item.replace(orderedHeaderRegex,'$1 ');
            tmpItem = tmpItem.replace(headerRegex,(match,$1,$2) => {
                tmpHeader = $1;
                //排除 第一个标题不是最大的标题 和 最小标题不是h6的情况
                if(tmpHeader.length < firstHeader.length || tmpHeader.length > HEADER_LEN){
                    return tmpItem;
                }else{
                    //当前header对应的位置
                    let targetIndex = tmpHeader.length - firstHeader.length;
                    headerList[targetIndex]++;
                    //重置所有子集的序号
                    let subSetLen = HEADER_LEN - tmpHeader.length + 1;
                    headerList.splice(targetIndex + 1,subSetLen,...(new Array(subSetLen)).fill(0));

                    let targetHeader = headerList.slice(0, targetIndex + 1);

                    return `${$1} ${targetHeader.join('.')} ${$2}`;
                }

            });

            return tmpItem;
        });

        return resultLines.join(LINE_SEP);
    },
    findFirstHeader:(fileLines = []) => {
        let firstHeader = '';
        let tmpResult;
        fileLines.some((item) => {
            tmpResult = item.match(/^(#+)\s*/);
            if(tmpResult){
                firstHeader = tmpResult[1]
            }
            return tmpResult;
        });
        return firstHeader;
    }
};


export default HeaderOrder
export {
    HeaderOrder
}